Task 2.2

Mitoschka opened the pull request.

Mitoschka committed:
Add Task 2.2

Mitoschka committed:
Update .yml

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+        /// </summary>
+        /// <param name="degree">Degree of received number</param>
+        /// <returns>Returns the power of a number</returns>
+        public int PowerOfTwo(int degree)
```
Зачем функция возведения в степень находится прямо в хеш-таблице, а не в отдельном статическом классе? Чем библиотечная плоха? Почему она не static? Почему она public? :)

yurii-litvinov commented on Task 2.2/Task 2.2.Test/MyHash.Test.cs:
```diff
@@ -0,0 +1,42 @@
+        [Test]
+        public void PowerOfTwoShouldWork()
+        {
+            HashTable hashTable = new HashTable();
```
var

yurii-litvinov commented on Task 2.2/Task 2.2.Test/MyHash.Test.cs:
```diff
@@ -0,0 +1,42 @@
+        [Test]
+        public void AddElementToHashTableAndDeleteElementOfHashTableShouldWork()
+        {
+            HashTable hashTable = new HashTable();
```
Тоже над овынести в Setup()

yurii-litvinov commented on Task 2.2/Task 2.2.Test/MyHash.Test.cs:
```diff
@@ -0,0 +1,42 @@
+
+            Assert.IsFalse(hashTable.IsContainInHashTable("abc"));
+        }
+    }
```
Это и все тесты? А попытка удаления несуществующего элемента, добавления нескольких одинаковых элементов, проверка, что удаление одного элемента не приводит к удалению всех остальных, а что будет, если сразу тыщу элементов добавить?

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        const int capacity = 100;
+        MyList[] hashTable = new MyList[capacity];
```
private

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+        /// </summary>
+        /// <param name="value">Input value</param>
+        /// <returns>Returns result</returns>
+        public int HashFunction(string value)
```
Хеш-функия --- это внутреннее дело хеш-таблицы, она должна быть private

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+        public void DeleteElementOfHashTable(string value)
+        {
+            int index = HashFunction(value);
+            if (IsContainInHashTable(value) == true)
```
Не надо сравниваться с булевыми константами

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+            if (IsContainInHashTable(value) == true)
+            {
+                hashTable[index].DeleteElement(value);
+                Console.WriteLine($"\"{value}\" было удалено" + "\n");
```
Нельзя выводить ничего на консоль нигде, кроме, быть может, класса Program. Иначе такой класс нельзя будет использовать в ситуации, когда вывод на консоль нежелателен

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,123 @@
+            if (currentElement.value != value)
+            {
+                currentElement.next = new ListElement();
+                currentElement.next.value = value;
```
Имеет смысл ListElement сделатиь конструктор, который бы принимал value

yurii-litvinov replied:
Напоминаю :) Иначе копипаст

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,123 @@
+        public int SizeOfList()
+        {
+            return counter;
+        }
```
Лучше через =>

yurii-litvinov commented on Task 2.2/Task 2.2/Program.cs:
```diff
@@ -0,0 +1,98 @@
+            "'2' - Удалить значение в хеш-таблице \n" +
+            "'3' - Проверить значение на принадлежность \n" +
+            "'4' - Вызвать меню команд \n");
+        }
```
Нужна пустая строчка

yurii-litvinov commented on Task 2.2/Task 2.2/Program.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+        }
+    }
+}
```
Вообще, если есть тесты, Main можно не писать

Mitoschka committed:
Change Task 2.2

Mitoschka committed:
Merge branch 'master' into Task-2.2

yurii-litvinov commented on Task 2.2/Task 2.2.Test/MyList.Test.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace Task_2._2.Test
+{
+    public class Test
```
Есть Tests, а есть Test, Вам не кажется, что имена не очень?

yurii-litvinov commented on Task 2.2/Task 2.2/Degree.cs:
```diff
@@ -0,0 +1,24 @@
+        /// </summary>
+        /// <param name="degree">Degree of received number</param>
+        /// <returns>Returns the power of a number</returns>
+        public int PowerOfTwo(int degree)
```
static же, он не использует состояние объекта. И сам класс static, у него полей нет

yurii-litvinov commented on Task 2.2/Task 2.2/MyExeption.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Task_2._2
+{
+    [Serializable]
+    public class MyException : Exception
```
Имя не очень. Имя исключения должно говорить, что произошло

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,87 @@
+        /// <returns>Returns result</returns>
+        private int HashFunction(string value)
+        {
+            Degree degree = new Degree();
```
var? Но вообще, это должен быть static-класс :)

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,87 @@
+            else
+            {
+                return;
+            }
```
Не думаю, что эти строки хоть что-нибудь полезное делают.

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,87 @@
+            }
+            else
+            {
+                throw new MyException("Такого элемента здесь нет");
```
А если элемента нет, то значение HashFunction не пригодилось, и мы только зря его считали, электричество жгли.

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,122 @@
+{
+    public class MyList
+    {
+        class ListElement
```
private

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,122 @@
+        /// <param name="value">Value to add</param>
+        public void AddElement(string value)
+        {
+            ListElement newElement = new ListElement();
```
var

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,122 @@
+            while (currentElement.next != null)
+            {
+                currentElement = currentElement.next;
+            }
```
А важно добавлять именно в конец, или можно добавлять всегда в начало? А если важно, почему это делается за линейное время, а не за константное?

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,122 @@
+            ListElement currentElement = head;
+            if (currentElement == null)
+            {
+                Console.WriteLine("Нет начала списка");
```
Нельзя писать на консоль нигде, кроме специально отведённых для этого мест, бросайте исключение, Вы же умеете уже

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,122 @@
+        /// </summary>
+        /// <param name="value">Value entered</param>
+        /// <returns>Returns check result</returns>
+        public bool IsContain(string value)
```
`Contains` более по-английски

yurii-litvinov commented on Task 2.2/Task 2.2/Program.cs:
```diff
@@ -0,0 +1,99 @@
+{
+    class Program
+    {
+        static void PrintMenu()
```
private

yurii-litvinov commented on Task 2.2/Task 2.2/Program.cs:
```diff
@@ -0,0 +1,99 @@
+
+        static void Main()
+        {
+            HashTable hashTable = new HashTable();
```
var

yurii-litvinov commented on Task 2.2/Task 2.2/Program.cs:
```diff
@@ -0,0 +1,99 @@
+                            Console.Write("[ПРОВЕРЯЕМ ЗНАЧЕНИЕ НА ПРИНАДЛЕЖНОСТЬ...]" + "\n" + "\n");
+                            Console.Write("Введите значение: " + "\n");
+                            string value = (Console.ReadLine());
+                            if (hashTable.IsContainInHashTable(value) == true)
```
Не надо сравниваться с булевыми константами

Mitoschka committed:
Update Task

Mitoschka committed:
Fix MyList

Mitoschka committed:
Fix Test

Mitoschka committed:
Merge branch 'master' into Task-2.2

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,139 @@
+            }
+            if (head.next == null)
+            {
+                head.next = tail;
```
Ммм? В этот момент head = tail, то есть эта строчка заставляет head.next указывать на head. К счастью, следующая строчка всё исправит, и head.next будет указывать на newElement. Но Вы уверены, что Вы написали то, что хотели?

Mitoschka committed:
Update Task

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,138 @@
+            }
+            tail.next = newElement;
+            tail = newElement;
+            ++counter;
```
"Если head.next == null, ДелайТак(); иначе тоже ДелайТак();". Иллюзия выбора.

Mitoschka committed:
Update Task

Mitoschka committed:
Update Task

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,135 @@
+            ListElement currentElement = head;
+            if (head == null)
+            {
+                currentElement = tail;
```
Если head == null, то и  tail должен быть равен null, потому что было бы странно иметь пустой список с непустым хвостом.

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,135 @@
+            if (head.value == value)
+            {
+                head = currentElement.next;
+                --counter;
```
А вот тут стоило бы поддерживать инвариант "head и tail одновременно null или не null". И в других местах, где что-то делается со списком (добавление и удаление вершин) --- тоже.

yurii-litvinov replied:
Не поправлено. Удаляете последний элемент ---  tail продолжает на него указывать

yurii-litvinov replied:
Отлично. Теперь при удалении любого элемента из головы, даже для списка из нескольких элементов, tail обнуляется. Вы специально пытаетесь получить незачёт? :)

yurii-litvinov reviewed (changes requested):
Постарайтесь не выстреливать пачкой случайных исправлений через пару часов после получения замечаний, а подумать, аккуратно реализовать и **проверить** исправления. Иначе я начну подозревать, что Вы сдаёте домашки как генетический алгоритм --- генерируя рандомные мутации кода и используя мои комментарии как фитнес-функцию. И начну с этим бороться --- например, проверять Ваши решения не чаще раза в неделю.

Mitoschka committed:
1) Fixed the implementation of the sheet using the head and tail. 2) Added a couple of comments

Mitoschka committed:
Исправил инвариант. Добавил дополнительные проверки в тест

Mitoschka committed:
Added new tests

1) Attempt to delete an empty list
2) Attempt to delete an empty hash table

yurii-litvinov commented on Task 2.2/Task 2.2.Test/MyHash.Test.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void MyExceptionInDeleteElementOfHashTableShouldWork()
+        {
+            Assert.Throws<System.NullReferenceException>(() => hashTable.DeleteElementOfHashTable("abc"));
```
А вот не должно оно бросать NullReferenceException. NullReferenceException означает ошибку в коде.

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,85 @@
+            }
+            else
+            {
+                throw new NullReferenceException("Такого элемента здесь нет");
```
Нельзя :) Это либо ArgumentException, либо InvalidOperationException

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            var newElement = new ListElement(value)
+            {
+                value = value
```
Што? Мы ж его в конструктор передаём, зачем его снова выставлять тут?

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,141 @@
+            {
+                value = value
+            };
+            if ((head == null) && (tail == null))
```
Если head == null, то tail == null, это инвариант списка. Вторую часть условия можно не проверять

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,141 @@
+            ListElement currentElement = head;
+            if ((head == null) && (tail == null))
+            {
+                throw new NullReferenceException("Нет начала списка");
```
Нельзя :)

Mitoschka committed:
Fixed comments

Fixed NullReferenceException on InvalidOperationException
Removed an extra constructor call
Fixed invariant
I remove the tail when removing the head

Mitoschka committed:
Add new XML

Mitoschka committed:
Merge branch 'master' into Task-2.2

Mitoschka committed:
Delete "<param name="args">Program arguments.</param>"

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,88 @@
+﻿using System;
+
+/// <summary>
+/// Global namespace.
```
Комментарии к пространствам имён не пишут, иначе в каждом файле придётся их писать

yurii-litvinov replied:
Не поправлено :)

yurii-litvinov commented on Task 2.2/Task 2.2/MyHash.cs:
```diff
@@ -0,0 +1,88 @@
+            int result = 0;
+            for (int i = 0; i != value.Length; ++i)
+            {
+                result = (result + Degree.PowerOfTwo(i) * value[i]) % capacity;
```
Degree.PowerOfTwo(i) может стать отрицательным для достаточно больших i, и тогда всё упадёт

Mitoschka committed:
-Fixed tail zeroing -Fixed the ability to get a negative index

Mitoschka committed:
Fix Delete and Fix Hash

yurii-litvinov commented on Task 2.2/Task 2.2/MyList.cs:
```diff
@@ -0,0 +1,146 @@
+                    if (tail.value == value)
+                    {
+                        tail = head;
+                    }
```
![image](https://user-images.githubusercontent.com/706768/84184989-004c3780-aa97-11ea-837e-dfd6a96d5f52.png)

Если в списке значение в голове случайно совпадает со значением в хвосте, то список разрывает в кашу.

Mitoschka committed:
Fixed error of delete in list

Mitoschka committed:
Fixed one more error

yurii-litvinov reviewed (approved):
Да! Зачтена

Mitoschka committed:
Merge branch 'master' into Task-2.2
